### PR TITLE
Introduce database schemata

### PR DESCRIPTION
**Pull request by CraigFe**
Alternative to https://github.com/mirage/irmin/pull/1454.

### CODE REVIEW

**Comment by CraigFe** [comment]
Tests are passing here now.

Can probably do some more cleanup, and may extract the `Node` / `Commit` privacy change to a separate PR.

**Comment by CraigFe** [comment]
Now rebased over https://github.com/mirage/irmin/pull/1471 and passing the CI.

Reviews much appreciated :-)

File: src/irmin-pack/ext.ml [hunk 1]
```diff
@@ -17,29 +17,24 @@
 open! Import
 module IO = IO.Unix
 
-module Maker
-    (V : Version.S)
-    (Config : Conf.S)
-    (Node : Irmin.Node.Maker)
-    (Commit : Irmin.Commit.Maker) =
-struct
+module Maker (V : Version.S) (Config : Conf.S) = struct
```
Comment (Ngoguey42) [inline src/irmin-pack/ext.ml:20]: I was hoping that the `Maker` level would be collapsed by moving `V` and `Config` to `Schema`.

Is this doable?

**Review by Ngoguey42 — APPROVED** [review]
I like this patch a lot. It will make my life easier :) Thanks!

(I didn't inspect every bits)

File: src/irmin-pack/ext.ml [hunk 2]
```diff
@@ -17,29 +17,24 @@
...
-  module Make
-      (M : Irmin.Metadata.S)
-      (C : Irmin.Contents.S)
-      (P : Irmin.Path.S)
-      (B : Irmin.Branch.S)
-      (H : Irmin.Hash.S) =
-  struct
+
+  module Make (Schema : Irmin.Schema.S) = struct
+    open struct
```
Comment (icristescu) [inline src/irmin-pack/ext.ml:24]: nice :)

File: src/irmin/private.ml [hunk 3]
```diff
@@ -18,20 +18,33 @@ open! Import
...
   (** Private content store. *)
+  module Contents :
+    Contents.Store with type key = Hash.t and type value = Schema.Contents.t
 
-  module Node : Node.Store with type key = Hash.t
   (** Private node store. *)
+  module Node :
+    Node.Store
+      with type key = Hash.t
+      (* and type value = Schema.node *)
```
Comment (icristescu) [inline src/irmin/private.ml:34]: you forgot to remove these comments

File: src/irmin-fs/irmin_fs.ml [hunk 4]
```diff
@@ -436,3 +418,9 @@ module IO_mem = struct
     Hashtbl.clear t.watches;
     Lwt.return_unit
 end
+
+(* Enforce that {!S} is a sub-type of {!Irmin.Maker}. *)
+module _ : Irmin.Maker = Maker (IO_mem)
```
Comment (icristescu) [inline src/irmin-fs/irmin_fs.ml:423]: We should maybe not use anonymous modules if we want to support older versions than `ocaml.4.10`

File: src/irmin/schema.ml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  module Info = Info.Default
+  module Branch = Branch.String
+  module Commit = Commit.Make (Hash)
+  module Path = Path.String_list
+  module Metadata = Metadata.None
+  module Node = Node.Make (Hash) (Path) (Metadata)
+  module Contents = C
+end
+
+module Abstract (S : S) :
```
Comment (icristescu) [inline src/irmin/schema.ml:55]: I don't think you are using this module anywhere (but I might have missed it)

**Review by icristescu — APPROVED** [review]
Really neat, thanks!

File: src/irmin/schema.ml (hunk 5 above)
Comment (CraigFe) [reply src/irmin/schema.ml:55]: Nice spot, thanks.

File: src/irmin-pack/ext.ml (hunk 1 above)
Comment (CraigFe) [reply src/irmin-pack/ext.ml:20]: I had a go at this. I'm unsure as to whether it works out to be better overall -- neither the config nor the version impacts the API of the store, so there are a few call-sites in the code where we want to be generic over one or the other. Regardless, the diff is quite separate from this one so I'll package it in a second PR and we can discuss from there.

**Comment by CraigFe** [comment]
Thanks both for the reviews. Merging now.

**Pull request merged**